registry/backoff.go

```diff
@@ -0,0 +1,108 @@
+package registry
+
+import (
+	"errors"
+	"net/http"
+	"sync"
+	"time"
+)
+
+var (
+	// For testing
+	now = time.Now
+)
+
+type hostBackoffRoundTripper struct {
+	roundTripper http.RoundTripper
+	max          time.Duration
+	backoffs     map[string]*backoff
+	sync.Mutex
+}
+
+// HostRateLimitedRoundTripper is a http.RoundTripper which applies throttling
+// to requests on a per-host*credentials basis.
+// Note: There's no cleanup on host/cred tuples so this will slowly leak.
+//
+// r              -- upstream roundtripper
+// maxBackoff     -- maximum length to backoff to between request attempts
+func HostBackoffRoundTripper(r http.RoundTripper, maxBackoff time.Duration) http.RoundTripper {
+	return &hostBackoffRoundTripper{
+		roundTripper: r,
+		max:          maxBackoff,
+		backoffs:     map[string]*backoff{},
+	}
+}
+
+func (c *hostBackoffRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	key := backoffKey{host: r.URL.Host}
+	if user, pass, ok := r.BasicAuth(); ok {
+		key.user = user
+		key.pass = pass
+	}
+	c.Lock()
```

Comment:
Can you do that? Where is that lock coming from? The RoundTripper?

What are you locking here? The backoff map?

Might be worth using an external read/write mutex, so if it is used elsewhere (either in your code now, or in the future) it makes it a bit more obvious that the map read/write is not thread safe.